lib/srv/db/snowflake/engine.go

```diff
@@ -0,0 +1,642 @@
+	HTTPClient *http.Client
+	// tokens is a tokens cache that holds the current session and master token
+	// used in the communication with the Snowflake.
+	tokens tokenCache
+}
+
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.clientConn = clientConn
+	e.sessionCtx = sessionCtx
+	// Make sure that tokens are initialized. Tokens are not initialized in tests.
+	if e.tokens.tokens == nil {
+		e.tokens = newTokenCache()
+	}
+	return nil
+}
+
+func (e *Engine) SendError(err error) {
+	if e.clientConn == nil || err == nil || utils.IsOKNetworkError(err) {
+		return
+	}
+
+	// Assume internal server error HTTP 500 and override if possible.
+	statusCode := http.StatusInternalServerError
+	if trace.IsAccessDenied(err) {
+		statusCode = http.StatusUnauthorized
+	}
+
+	jsonBody := fmt.Sprintf(`{"success": false, "message:"%s"}`, err.Error())
+
+	response := &http.Response{
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		StatusCode: statusCode,
+		Body:       io.NopCloser(bytes.NewBufferString(jsonBody)),
+		Header: map[string][]string{
+			"Content-Type":   {"application/json"},
+			"Content-Length": {strconv.Itoa(len(jsonBody))},
+		},
+	}
+
+	dumpResponse, err := httputil.DumpResponse(response, true)
+	if err != nil {
+		e.Log.Errorf("snowflake error: %+v", trace.Unwrap(err))
+		return
+	}
+
+	_, err = e.clientConn.Write(dumpResponse)
+	if err != nil {
+		e.Log.Errorf("snowflake error: %+v", trace.Unwrap(err))
+		return
+	}
+}
+
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session) error {
+	accountName, err := extractAccountName(sessionCtx.Database.GetURI())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := e.authorizeConnection(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	defer e.Audit.OnSessionEnd(e.Context, sessionCtx)
+
+	clientConnReader := bufio.NewReader(e.clientConn)
+
+	for {
+		req, err := http.ReadRequest(clientConnReader)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		err = e.processRequest(ctx, sessionCtx, req, accountName)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// processRequest reads request from connected Snowflake client, processes the requests/responses and send data back
+// to the client.
+func (e *Engine) processRequest(ctx context.Context, sessionCtx *common.Session, req *http.Request, accountName string) error {
+	snowflakeToken, err := e.getConnectionToken(ctx, req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	requestBodyReader, err := e.processPath(ctx, req, accountName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	reqCopy, err := copyRequest(ctx, req, requestBodyReader)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	reqCopy.URL.Scheme = "https"
+	reqCopy.URL.Host = sessionCtx.Database.GetURI()
+
+	e.setAuthorizationHeader(reqCopy, snowflakeToken)
+
+	// Send the request to Snowflake API
+	resp, err := e.HTTPClient.Do(reqCopy)
```

Comment:
There are some corner cases like flushing, web socket handling in current implementation wonder if `httputil.ReverseProxy`  or `oxy/forward` can be leverage here to make a call to Snowflake API  upstream server ?